File: tests/wasm_wrapper_test.c
/* Integration test for the cwist-wasm JS wrapper (wasm/npm/index.js).
 *
 * Builds a consumer-style app module via include/cwist/wasm/wasm_entry.h,
 * then tests/wasm_wrapper_test.js drives it through the wrapper with no
 * Emscripten-specific code on the JS side. Run:
 *
 *   make wasm-wrapper-test
 *
 * clang-format off regions: EM_JS bodies are JS, not C (see typedarray.h). */
#include <cwist/sys/app/app.h>
#include <cwist/wasm/typedarray.h>
#include <cwist/wasm/wasm_entry.h>
#include <string.h>

static cwist_app *g_app;

CWIST_WASM_INSTALL_VIEWS()

static void hello_handler(cwist_http_request *req, cwist_http_response *res) {
    (void)req;
    cwist_sstring_assign(res->body, "hello-from-wrapper-test");
    cwist_http_header_add(&res->headers, "Content-Type", "text/plain");
    cwist_http_header_add(&res->headers, "X-Wrapper-Test", "yes");
}

static void echo_handler(cwist_http_request *req, cwist_http_response *res) {
    /* Echo the request body back so the JS side can verify body plumbing. */
    if (req->body && req->body->data) {
        cwist_sstring_assign(res->body, req->body->data);
    }
    cwist_http_header_add(&res->headers, "Content-Type", "application/octet-stream");
}

/* File scope: the macro defines functions, which C does not allow inside
 * another function body. g_app is populated in main() before any dispatch. */
CWIST_WASM_DEFINE_ENTRY(g_app)

int main(void) {
    g_app = cwist_app_create();
    if (!g_app) return 1;
    cwist_app_get(g_app, "/hello", hello_handler);
    cwist_app_post(g_app, "/echo", echo_handler);
    cwist_wasm_install_views();
    return 0;
}

File: tests/wasm_wrapper_test.js
/* Drives tests/wasm_wrapper_test.c (compiled to wrapper_test.js) through the
 * cwist-wasm wrapper and asserts the fetch-shaped API end to end. Run by
 * `make wasm-wrapper-test` after the module is built (MODULARIZE output:
 * the glue exports createCwistModule() instead of populating a global). */
'use strict';

const path = require('path');
const { createCwist } = require(path.join(__dirname, '..', 'wasm', 'npm'));
const createCwistModule = require(path.join(process.cwd(), 'wrapper_test.js'));

(async () => {
  const Module = await createCwistModule();
  const handle = createCwist(Module);

  /* 1. GET with no body: status, headers, payload all verified. */
  const res = handle({ method: 'GET', path: '/hello' });
if (res.status !== 200) throw new Error('status: ' + res.status);
if (res.headers['Content-Type'] !== 'text/plain') {
  throw new Error('Content-Type: ' + res.headers['Content-Type']);
}
if (res.headers['X-Wrapper-Test'] !== 'yes') {
  throw new Error('X-Wrapper-Test missing: ' + JSON.stringify(res.headers));
}
const text = new TextDecoder().decode(res.body);
if (text !== 'hello-from-wrapper-test') throw new Error('body: ' + JSON.stringify(text));

/* 2. POST with a Uint8Array body: request body plumbing into C handlers.
 * Avoid a leading NUL: the echo handler uses cwist_sstring_assign, which
 * is strlen-based. */
const payload = new Uint8Array([0x01, 0x02, 0xfe, 0xff]);
const echoed = handle({ method: 'POST', path: '/echo', body: payload });
if (echoed.status !== 200) throw new Error('echo status: ' + echoed.status);
if (echoed.body.length !== payload.length ||
    !payload.every((b, i) => echoed.body[i] === b)) {
  throw new Error('echo body mismatch: ' + Array.from(echoed.body));
}

/* 3. Response body must be a copy: mutate it, dispatch again, expect the
 * original bytes. */
res.body[0] = 0x58;
const again = handle({ method: 'GET', path: '/hello' });
if (new TextDecoder().decode(again.body) !== 'hello-from-wrapper-test') {
  throw new Error('body view aliasing detected');
}

console.log('wasm_wrapper_test: OK (GET, POST echo, body-copy isolation)');
})().catch((err) => {
  console.error(err);
  process.exit(1);
});
